packages/tpetra/core/src/Tpetra_FECrsMatrix_def.hpp

```diff
@@ -0,0 +1,163 @@
+#define TPETRA_FECRSMATRIX_DEF_HPP
+
+#include "Tpetra_CrsMatrix.hpp"
+
+namespace Tpetra {
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+FECrsMatrix(const Teuchos::RCP<const fe_crs_graph_type>& graph,
+            const Teuchos::RCP<Teuchos::ParameterList>& params) : 
+  // We want the OWNED_PLUS_SHARED graph here
+  // NOTE: The casts below are terrible, but necesssary
+  crs_matrix_type( graph->inactiveCrsGraph_.is_null() ? Teuchos::rcp_const_cast<crs_graph_type>(Teuchos::rcp_dynamic_cast<const crs_graph_type>(graph)) : graph->inactiveCrsGraph_,params),
+  feGraph_(graph)
+
+{
+  const char tfecfFuncName[] = "FECrsMatrix(RCP<const FECrsGraph>[, RCP<ParameterList>]): ";
+  typedef typename local_matrix_type::values_type values_type;
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
+    (graph.is_null (), std::runtime_error, "Input graph is null.");
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
+    (!graph->isFillComplete (), std::runtime_error, "Input graph is not "
+     "fill complete. You must call fillComplete on the graph before using "
+     "it to construct a FECrsMatrix.  Note that calling resumeFill on the "
+     "graph makes it not fill complete, even if you had previously called "
+     "fillComplete.  In that case, you must call fillComplete on the graph "
+     "again.");
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
+     ( *graph->activeCrsGraph_!= fe_crs_graph_type::FE_ACTIVE_OWNED,std::runtime_error,
+      "Input graph must be in FE_ACTIVE_OWNED mode when this constructor is called.");
+
+  activeCrsMatrix_     = Teuchos::rcp(new FEWhichActive(FE_ACTIVE_OWNED_PLUS_SHARED));
+
+  // Make an "inactive" matrix, if we need to
+  if(!graph->inactiveCrsGraph_.is_null() ) {
+    // We are *requiring* memory aliasing here, so we'll grab the first chunk of the Owned+Shared matrix's values array to make the 
+    // guy for the Owned matrix.
+    values_type myvals = this->getLocalMatrix().values;
+
+    size_t numOwnedVals = graph->getLocalGraph().entries.extent(0); // OwnedVals
+    inactiveCrsMatrix_ = Teuchos::rcp(new crs_matrix_type(graph,Kokkos::subview(myvals,Kokkos::pair<size_t,size_t>(0,numOwnedVals))));
+  }
+}
+
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>&
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+operator=(const FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>& rhs)
+{
+  return *this;
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::doOwnedPlusSharedToOwned(const CombineMode CM) {
+  if(!inactiveCrsMatrix_.is_null() && *activeCrsMatrix_ == FE_ACTIVE_OWNED_PLUS_SHARED) {
+    // Do a self-export in "restricted mode"
+    this->doExport(*this,*feGraph_->importer_,CM,true);
+    inactiveCrsMatrix_->fillComplete();
```

Comment:
Do you need to pass this a domain and range Map?

Reply:
No.  The graph already has those.